### PR TITLE
Generalize handshake protocol, using TLS as a specific example.

### PR DESCRIPTION
**Pull request by chris-wood**

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1493,8 +1493,8 @@ supports.  The version field in the QUIC packet header is authenticated using
 transport parameters.  The position and the format of the version fields in
 transport parameters MUST either be identical across different QUIC versions, or
 be unambiguously different to ensure no confusion about their interpretation.
-One way that a new format could be introduced is to define a TLS extension with
-a different codepoint.
+When using TLS as the handshake protocol, defining a new TLS extension with a
```
Comment (mimoo) [inline draft-ietf-quic-transport.md]: two things:

* If we're really writing an agnostic-to-TLS document then I think we can remove this sentence all together.
* In order to be able to swap the QUIC-TLS doc, we should provide a clear API ( -> we pass these transport arguments, <- the security layer gives them back) and clear security properties for the transport layer (it should authenticate the transport parameters)

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2348,8 +2347,8 @@ have negotiated a packet protection scheme with a larger minimum AEAD expansion.
 
 An endpoint SHOULD NOT send a stateless reset that is significantly larger than
 the packet it receives.  Endpoints MUST discard packets that are too small to be
-valid QUIC packets.  With the set of AEAD functions defined in {{QUIC-TLS}},
-packets that are smaller than 21 bytes are never valid.
+valid QUIC packets.  When using TLS as the handshake protocol, no AEAD
+function defined in {{QUIC-TLS}} will yield packets smaller than 21 bytes.
```
Comment (mimoo) [inline draft-ietf-quic-transport.md]: again, not great. What should other ~transport layers~~ handshake protocol do? What

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -2575,9 +2574,11 @@ to have confidentiality and data origin authentication; the cryptographic
...
 corresponding keys.
 
-The packet number field contains a packet number, which has additional
-confidentiality protection that is applied after packet protection is applied
-(see {{QUIC-TLS}} for details).  The underlying packet number increases with
+The packet number field contains a packet number. Endpoints use these numbers
+to protect individual packets. After applying packet protection, packet numbers
+are themselves protected encryption. TLS-based packet number protection is
+described in see {{QUIC-TLS}} for details. Future versions may protect packet
+numbers in a different manner. The underlying packet number increases with
```
Comment (mimoo) [inline draft-ietf-quic-transport.md]: this one is tricky, as it is deeply entangled with the QUIC-TLS RFC

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -2629,7 +2630,8 @@ Packet numbers are divided into 3 spaces in QUIC:
 - Application data space: All 0-RTT and 1-RTT encrypted packets
   {{packet-protected}} are in this space.
 
-As described in {{QUIC-TLS}}, each packet type uses different protection keys.
+Each packet type uses different protection keys; see {{QUIC-TLS}} for a description
```
Comment (mimoo) [inline draft-ietf-quic-transport.md]: As I understand, TLS gives us back different keys for different types of packets, and it's this RFC's job to encrypt stuff afterwards. If my understanding is correct, then we should probably list the type of keys here, and have a clear API as to how we retrieve these keys from the handshake protocol.

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -2655,9 +2657,12 @@ it receives.
...
 packet number space. Duplicate suppression MUST happen after removing packet
-protection for the reasons described in Section 9.3 of {{QUIC-TLS}}. An
-efficient algorithm for duplicate suppression can be found in Section 3.4.3 of
-{{?RFC4303}}.
+protection. Otherwise, an attacker may guess values for packet numbers and
+have an endpoint confirm guesses through timing side channels. For authentication
+to be free from side-channels, the entire process of packet number protection removal,
+packet number recovery, and packet protection removal MUST be applied together
+without timing and other side-channels. An efficient algorithm for duplicate suppression
+can be found in Section 3.4.3 of {{?RFC4303}}.
```
Comment (mimoo) [inline draft-ietf-quic-transport.md]: this is good

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -3354,18 +3360,20 @@ Long Packet Type (T):
...
 
 Packet Number Length (P):
 
 : The least significant two bits (those with a mask of 0x03) of byte 0 contain
   the length of the packet number, encoded as an unsigned, two-bit integer that
   is one less than the length of the packet number field in bytes.  That is, the
   length of the packet number field is the value of this field, plus one.  These
-  bits are protected using header protection (see Section 5.4 of {{QUIC-TLS}}).
+  bits are protected using header protection (see Section 5.4 of {{QUIC-TLS}}
+  for an example using TLS).
```
Comment (mimoo) [inline draft-ietf-quic-transport.md]: I don't really like that TLS cannot be used as a blackbox and modifies part of the QUIC packets itself.

File: draft-ietf-quic-transport.md (hunk 4 above)
Comment (chris-wood) [reply draft-ietf-quic-transport.md]: No, not quite. Packet protection is described in the TLS doc, not here.

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -5040,7 +5050,8 @@ CRYPTO_ERROR (0x1XX):
 : The cryptographic handshake failed.  A range of 256 values is reserved for
   carrying error codes specific to the cryptographic handshake that is used.
   Codes for errors occurring when TLS is used for the crypto handshake are
-  described in Section 4.8 of {{QUIC-TLS}}.
+  described in Section 4.8 of {{QUIC-TLS}}. Future QUIC versions SHOULD define
+  separate error CRYPTO_ERROR codes.
```
Comment (mimoo) [inline draft-ietf-quic-transport.md]: did you mean "separate CRYPTO_ERROR error codes" ?

Comment (chris-wood) [reply draft-ietf-quic-transport.md]: Yeah, will fix.

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (chris-wood) [reply draft-ietf-quic-transport.md]: We can add something that says "other handshake protocols MUST not yield packets less than 21 bytes." Thoughts?

**Pull request closed without merging**